### ISSUE TITLE
Can't re-initialise TestObservationRegistry between tests

### ISSUE THREAD

**Issue by wabrit** [micrometer-metrics/micrometer#3674]
Apologies if I am missing something obvious but I'm failing to see a way to "reuse" a TestObservationRegistry across multiple tests in a test class.

Example:

```java
@ExtendWith(SpringExtension.class)
@SpringBootTest
public class FooServiceTest {
  @Autowired
  private FooService fooService;

  @Autowired
  private TestObservationRegistry observationRegistry;

  @TestConfiguration
  static class ObservationConfiguration {
    @Bean
    TestObservationRegistry observationRegistry() {
      return TestObservationRegistry.create();
    }
  }

  @Test
  void testAddFooWithABar() {
    fooService.addWithBar();
    TestObservationRegistryAssert.assertThat(observationRegistry)
        .hasObservationWithNameEqualTo("foo.create")
        .that()
        .hasLowCardinalityKeyValue("isBar", "true");
  }

  @Test
  void testAddFooWithoutABar() {
    fooService.addWithoutBar();
    TestObservationRegistryAssert.assertThat(observationRegistry)
        .hasObservationWithNameEqualTo("foo.create")
        .that()
        .hasLowCardinalityKeyValue("isBar", "false");
  }
}
```

In the above, both tests call methods of the FooService which create similar Observations but with a slightly different KeyValue `isBar`.

When I run the test, one of the tests always fails because the `hasObservationWithNameEqualTo()` assertion gets the latest Observation matching the name `foo.create`, and the subsequent assertion on key value fails.

What is needed (I think) is some way of "clearing"/re-initialising the `TestObservationRegistry` between tests, but I can't see any way of doing this short of annotating each test method with `@DirtiesContext`, which seems like a sledgehammer to crack a nut, and slows down test runs. 

Or am I going about this the wrong way?

**Comment by jonatan-ivanov** [comment]
It seems we don't have support for this scenario but I think what you are doing is valid, we have a couple of options (not mutually exclusive):
- Add a `clear` method (it might be useful for other scenarios too)
- Add a `hasObservationWithNameAndKeyValuesEqualTo(...)` method (I'm not sure this is a good idea)
- Fix the `That` class to accept a collection of contexts instead of one and let the rest of the asserts work on the collection instead of the first element (this might not be needed if we add a `clear` method)

Right now, as a workaround you can call `getCurrentObservation` and do these verifications on the context.

**Comment by wabrit** [comment]
Thanks for getting back so quickly - it feels to me like a clear() method would be good, because its a common paradigm for these things,  and it promotes isolation between tests.

**Comment by jonatan-ivanov** [comment]
I think there would also be the easiest fix, do you want to open a PR for it?
It seems it is as simple as calling `contexts.clear();` and adding `clear` methods + some javadoc.

**Comment by wabrit** [comment]
Happy to give it a try - can you point at any relevant guidlines for the project for coding/process to make sure I do things the right way (also do I need permission to create a new branch?). TIA

**Comment by bclozel** [comment]
I think that in a typical unit test setup, the `TestObservationRegistry`, as a test class field, will be newly instantiated for each test. This is only a problem with integration tests and an actual application context where a single registry instance is expected for the running server.

In this case, adding a `clear()` method could help if it's used with `@BeforeEach` to ensure that the registry is empty. If developers enable parallel test execution this becomes a problem again and they would need to worry about shared state and probably use `@ResourceLock`. Since the application context is itself cached by the test context framework, the context (and the `TestObservationRegistry` bean) can be reused by another test class and side effects can also happen there.

In short, I'm not sure this really solves the problem as by definition the registry is stateful within a live application.

**Comment by wabrit** [comment]
Hi @bclozel does that imply that I'm driving this wrong in my tests? It doesn't seem to be newly instantiated per test in my test class from my own observations, which is why I had to use `@DirtiesContext` on any test that used it to force reinstantation and preserve isolation.

The approach seems very similar to e.g. `TestChannelBinderConfiguration` in Spring Cloud; that provides you with various beans that have to be cleared between tests, and expose a `clear()` method for this purpose.

**Comment by bclozel** [comment]
@wabrit you're not doing anything wrong - as stated in my previous comment, integration tests like this expect a running application context that's shared between test methods. I guess what I'm saying is that for "regular" unit tests the registry would be created each time, like:

```java
public class MyTests {

  private TestObservationRegistry registry = TestObservationRegistry.create();

  @Test
  void firstTest() {
    // a new MyTests instance is created for running each @Test method
  }

  @Test
  void secondTest() {
    // a new MyTests instance is created for running each @Test method
  }

}
```

I'm not familiar with the `TestChannelBinderConfiguration` approach. I'm merely pointing out that exposing a `clear()` method is not enough to isolate integration tests like this.

**Comment by wabrit** [comment]
Thanks @bclozel - I guess I didn't think of this as an "integration test" , but probably that's a minor point - it feels like a unit test that uses a particular customization of the registry which therefore needs to be "isolated" from intra-test coupling.

Am I right in thinking that implicit in your example here (cf. the approach of using `@TestConfiguration` to create a bean for the registry, which is what is done in the samples I've googled so far) is that the registry needs to be passed down into the beans that create the observations (`FooService` in my example) above explicitly via e,g, a constructor? 

I was trying to avoid that approach as I'm applying observations to a large number of projects that don't use constructor autowiring and that's going to be a huge code change (unless I've misunderstood what needs to be done). Which was why `clear()` seemed an attractive option in this case,

**Comment by bclozel** [comment]
As soon as you're using `@SpringBootTest` or any test context support that creates an application context, you're expecting beans to be injected and broader Spring support. This approach is perfectly fine. You shouldn't have to rewrite your tests to not use this mechanism.

**Comment by wabrit** [comment]
FYI - in the meantime as a workaround (since cloning the micrometer source onto a Windows box seems to result in build issues which need some unpicking), and as a confirmation of your suggestion @jonatan-ivanov, I implemented a simple wrapper which uses reflection to clear the contexts on a `TestObservationRegistry`. That allowed me to dispense with the use of `@DirtiesContext` in my Spring Boot unit tests.

**Comment by marcingrzejszczak** [comment]
With this PR https://github.com/micrometer-metrics/micrometer/pull/3679 you will be able to provide a custom assertion where you will provide all the assertions you want and if at least one observation passes the assertion then you're ok. In you case you would give it a name and key values.

**Comment by wabrit** [comment]
Hi @marcingrzejszczak thanks for the info; it didn't appear to me that this addresses the requirement of test isolation,  but i might be missing something.

**Comment by marcingrzejszczak** [comment]
The test isolation doesn't matter in this case because it's the same case as if you had a shared in memory database (i'm referring to what @bclozel mentioned about the Spring context is shared between tests). You can either clear it every single time but your tests won't be then thread safe or write your assertions in such a way that they don't care about other entries but are pinpointing exactly what you want to assert. Of course I can add the `clear` method but I don't think it's a good idea.

**Comment by wabrit** [comment]
OK thanks all - I think basically we're on slightly different pages; I'm dealing with a large pre-existing set of microservices and unit tests, to which I am adding observability assertions, that are not thread-safe (they run in series not parallel) and therefore have no qualms about reinitializing state between tests; it seemed like lots of other libraries adopted the `clear()` paradigm to assist in this.

I understand your approach with the `hasAnObservation` assertion - I'm just not sure that it isolates test _code_ in quite the same way as `clear()` does; unless I've missed something each test needs a sufficiently unique set of observation criteria (keys, etc) in order for assertions not to get "mixed up" (kind of the flipside to my original scenario) i.e. I _can_ focus on what observations I am interested in, but I also have to be mindful of what observations other tests might have triggered because they might not be unique.  

So for me it doesn't seem quite as bullet-proof as `clear()` would be in my non-parallel testing world, but I can understand if you're loathe to support that if you think its going to be misused in some way.

**Comment by wabrit** [comment]
Thanks very much @marcingrzejszczak and for your collective patience and helpfulness everyone - now I'm feeling guilty and a little inadequate that my tests are not as they should be :-)  In that spirit I will try and pick a sample test class and see if I can parallel it up successfully, which will then at least give me the option of using your recommended approach.

**Comment by marcingrzejszczak** [comment]
Don't be so hard on yourself ;) That's a typical problem when running tests against a shared state. Either you lock on the state and ensure that for each of the tests from the beginning till the end you have control of it which leads of course to lower performance, or you ignore the fact that you have other tests injecting entries to your shared state but you create test data in such a way that you can later properly query them.

Example: you create a user with a name that is e.g. timestamp or name of your test + timestamp and then when you inject it to the in memory database you test that that concrete entry is there in the database. You don't check that the database size == 1 cause in the meantime some other test could have entered something to your db.
